File: Js/ReservationService.js
const Reservation = require('./reservation');

class ReservationService {
    #reservations = [];

    estDisponible(objet, date) {
        return !this.#reservations.some(
            (r) =>
                r.objet.nom.toLowerCase() === objet.nom.toLowerCase() &&
                r.date.toISOString().split('T')[0] === date.toISOString().split('T')[0]
        );
    }

    reserverObjet(objet, date) {
        if (!this.estDisponible(objet, date)) {
            throw new Error(`L'objet '${objet.nom}' est déjà réservé le ${date.toISOString().split('T')[0]}`);
        }

        this.#reservations.push(new Reservation(objet, date));
        console.log(`✅ Réservation confirmée pour '${objet.nom}' le ${date.toISOString().split('T')[0]}`);
    }
}

module.exports = ReservationService;

File: Python/reservation_service.py
from datetime import date
from typing import List
from objet_louable import ObjetLouable
from reservation import Reservation

class ReservationService:
    def __init__(self):
        self._reservations: List[Reservation] = []

    def est_disponible(self, objet: ObjetLouable, date: date) -> bool:
        return not any(
            r.objet.nom.lower() == objet.nom.lower() and r.date == date
            for r in self._reservations
        )

    def reserver_objet(self, objet: ObjetLouable, date: date):
        if not self.est_disponible(objet, date):
            raise ValueError(f"L'objet '{objet.nom}' est déjà réservé le {date}")
        
        self._reservations.append(Reservation(objet, date))
        print(f"✅ Réservation confirmée pour '{objet.nom}' le {date}")